### PR TITLE
Integrate ModSecurity with AzWAF

### PR DESCRIPTION
**Pull request by vkrivopalov**
## Description

This fix makes ModSecurity pass on the request if it has previously been
allowed by AzWAF.

AzWAF processing takes place before ModSecurity and as such may result
in request being entirely whitelisted by a custom rule.
In this case we should forward the request without having ModSecurity
process it to avoid accidentally blocking it.

## Testing

Tested with a Onebox test that configures a custom rule which allows requests containing `1` in a query parameter and sends a request to http://localhost/?a=1=1. This request is normally blocked by ModSecurity with CRS but here results in 200 OK response as ModSecurity is bypassed according to AzWAF custom rules processing result.

### CODE REVIEW

File: nginx/modsecurity/ngx_http_modsecurity.c [hunk 1]
```diff
@@ -635,6 +648,10 @@ ngx_http_modsecurity_init(ngx_conf_t *cf)
     }
 #endif    
 
+    azwaf_processing_result_key = ngx_hash_key_lc(
```
Comment (azhao155) [inline nginx/modsecurity/ngx_http_modsecurity.c:651]: what's the usage of the key here, just to look up the variable when receive the request?

File: nginx/modsecurity/ngx_http_modsecurity.c [hunk 2]
```diff
@@ -854,6 +871,20 @@ ngx_http_modsecurity_handler(ngx_http_request_t *r)
...
+    ngx_http_variable_value_t* azwaf_processing_result =
+        ngx_http_get_variable(r, &azwaf_processing_result_var, azwaf_processing_result_key);
+    if (azwaf_processing_result != NULL && !azwaf_processing_result->not_found) {
+        if (ngx_strncasecmp(
+                azwaf_processing_result->data,
+                azwaf_action_allow.data,
+                MIN(azwaf_processing_result->len ,azwaf_action_allow.len)) == 0) {
+
+            ngx_log_error(NGX_LOG_INFO, r->connection->log, 0, "The request is allowed by AzWAF, skip ModSecurity processing");
+            return NGX_DECLINED;
```
Comment (azhao155) [inline nginx/modsecurity/ngx_http_modsecurity.c:884]: so decline is to skip the module?

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 1 above)
Comment (vkrivopalov) [reply nginx/modsecurity/ngx_http_modsecurity.c:651]: Yes, it is required by the lookup function.

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 2 above)
Comment (vkrivopalov) [reply nginx/modsecurity/ngx_http_modsecurity.c:884]: Right, the name is quite misleading but effectively `NGX_DECLINED` means `move on to the next module in chain`.

**Review by azhao155 — APPROVED** [review]

**Pull request merged**